### PR TITLE
Added ETL data pipeline

### PR DESCRIPTION
**Pull request by aybanda**
@bellanov 
## Summary
Enhanced ETL pipeline in main.py with improved data handling, error management, and additional processing steps.

## Changes Made
- Implemented concurrent data extraction using ThreadPoolExecutor
- Added data_entry function for manual data input
- Introduced quality_assurance function to remove duplicates and ensure data integrity
- Enhanced error handling and logging throughout the pipeline
- Updated transform function to handle age data more robustly

## Testing
- Ran the updated ETL pipeline locally
- Verified successful extraction from both dummy data and external API
- Confirmed proper transformation, including new birth_year calculation
- Tested manual data entry and quality assurance processes
- Checked error handling by simulating various failure scenarios

Screenshot 2024-09-06 at 5 49 54 PM

## Performance Impact
- Improved extraction speed through concurrent processing

## Additional Notes
- The quality_assurance function now converts dictionary columns to strings before deduplication
- The pipeline now includes both CSV and JSON output formats

Buy me a coffee: https://buymeacoffee.com/aybanda

### CODE REVIEW

**Comment by aybanda** [comment]
@bellanov 
If you like my work and want to support me!

Buy me a coffee: https://buymeacoffee.com/aybanda

let me know if you have any issues!

**Comment by bellanov** [comment]
Will do! Thanks for your contribution!

**Comment by bellanov** [comment]
I've looked over your contribution and am pretty impressed with the implementation. It shows an excellent understanding of not only the task but its execution.

Barring some extraneous things like documentation, linting, organizing files, etc., I'm very satisfied with the changes and will pay you for them.

I'm very interested in collaborating further, as I have a lot to do and could use any capable help I can find :)

Let me know if you'd like to continue working together. You're also helping me test workflows for this task management solution :)

I'm currently setting up my own "Buy Me Coffee" account so I can get you some coffee!!!

**Comment by aybanda** [comment]
@bellanov feel free to reach me jsaj2024@gmail.com
I would love to contribute and collaborate with you.
Really appreciate your tip, thank you for that.

File: .DS_Store
Comment (bellanov) [inline .DS_Store:1]: Unnecessary file. Adding a .gitignore helps prevent unwanted files from committed to the repository.

Comment (bellanov) [reply .DS_Store:1]: A valid **.gitignore** for both Python and MacOS can be generated at gitignore.io

Add the content at the following link to a file called **.gitignore**. Then delete _.DS_Store_.

https://www.toptal.com/developers/gitignore/api/python,macos

The mac option covers all system files.

File: main.py
Comment (bellanov) [inline main.py:1]: This is a great start and we can extend this to 

**Ideas:**
- graduate from dummy data to accepting local file inputs. 
 - not every user will require an external API in their use case
 - linting (tslint / pylint)
 - testing (pytest)

File: main.py [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+    {"id": 3, "name": "Charlie", "age": 35},
+]
+
+def extract_dummy_data() -> pd.DataFrame:
+    logging.info("Extracting dummy data")
+    return pd.DataFrame(dummy_data)
+
+def extract_api_data() -> pd.DataFrame:
+    logging.info("Extracting data from external API")
+    response = requests.get("https://jsonplaceholder.typicode.com/users")
```
Comment (bellanov) [inline main.py:21]: Turning this into an ENVIRONMENT VARIABLE such that it can be more dynamic.

File: main.py [hunk 2]
```diff
@@ -0,0 +1,93 @@
+import pandas as pd
+import logging
+import requests
+from concurrent.futures import ThreadPoolExecutor
+
+logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(levelname)s - %(message)s')
+
+# Dummy data
+dummy_data = [
```
Comment (bellanov) [inline main.py:9]: Could take this one step further and read in information from local file(s), an input directory for instance.

File: main.py [hunk 3]
```diff
@@ -0,0 +1,93 @@
+import pandas as pd
```
Comment (bellanov) [inline main.py:1]: Need to have dependencies saved in requirements.txt . I had to install them myself to execute your code.

To isolate your dependencies
pip freeze > requirements.txt

**Review by bellanov — CHANGES_REQUESTED** [review]
Just a lot of tiny, straightforward tasks to get through on the road to turning this into a product, or something anyone can consume.

**Review by bellanov — APPROVED** [review]

**Comment by bellanov** [comment]
I've added you to the Andy project as a member, so I can assign other form of works to you in the future. 

Can use the above feedback as a test, you'll need to accept that invite so I can assign tasks to you.

**Pull request merged**